README.md
coxphSGD
================

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/coxphSGD)](http://cran.r-project.org/web/packages/coxphSGD) [![Total Downloads](http://cranlogs.r-pkg.org/badges/grand-total/coxphSGD?color=orange)](http://cranlogs.r-pkg.org/badges/grand-total/coxphSGD) [![Build Status](https://api.travis-ci.org/MarcinKosinski/coxphSGD.png)](https://travis-ci.org/MarcinKosinski/coxphSGD)

-   [Overview](#overview)
    -   [Applications](#applications)
    -   [Installation](#installation)
    -   [Assumptions](#assumptions)
-   [Simple example](#simple-example)
    -   [Simulated data](#simulated-data)
    -   [Fit the model](#fit-the-model)
    -   [Track estimates during each iteration on the contour lines plot](#track-estimates-during-each-iteration-on-the-contour-lines-plot)
    -   [Used code](#used-code)
-   [Genomic data analysis](#genomic-data-analysis)
    -   [The Cancer Genome Atlas Data and RTCGA](#the-cancer-genome-atlas-and-rtcga)
    -   [Fit the complex model](#fit-the-complex-model)
    -   [Estimated survival curves](#estimated-survival-curves)
-   [Mathematical formulas](#mathematical-formulas)
-   [useR 2017](#user-2017)
    -   [Long abstract](https://github.com/MarcinKosinski/useR2017/blob/master/abstract_MarcinKosinski.pdf)
    -   [Short abstract](https://github.com/MarcinKosinski/useR2017/blob/master/abstract_MarcinKosinski_short.pdf)
    -   [Presentation](#presentation)

Overview
========

Applications
------------

Know the `survival::coxph()` function that calculates the estimates of the Cox Proportional Hazards model? <br> It uses the **gradient descent order II** method (known also as Newton-Raphson method) to optimize the **partial** log-likelihood function of the Cox PH model.

The `coxphSGD::coxphSGD()` is an equivalent that uses the **stochastic gradient descent order I** method in the optimization process, which can be beneficial in situations like

-   the data is in a **streaming structure** and appear in batches
-   the data is of the great volume (computations and RAM issues) and one can't simply use all data for computations and needs to go **block by block** of the data

> The stochastic gradient descent order I method can be used with already calculated estimates to generate updated ones in the situation in which new data appear and one can't recalculate the whole model with all historical data.

Installation
------------

``` r
install.packages('coxphSGD') # once it is on CRAN
devtools::install_github('MarcinKosinski/coxphSGD') # development version
```

Usage.

``` r
library(coxphSGD)
## help(package = 'coxphSGD') # manual reference
```

Assumptions
-----------

The Cox Proportional Hazards model assumes (mainly) that the explanatory variables are constant over time and that the **hazard** (one of the survival measures) estimates between considered groups are proportional over the time.

This model calculates the estimates that can be interpreted as proportion of the hazard change while moving from one level of the considered explanatory variable to another (while all other variables are fixed) - explained with details in the [Mathematical formulas](#mathematical-formulas) section.

Simple example
==============

Simulated data
--------------

Let's simulate artificial/fake data that follows the Cox Proportional Hazards model assumptions. Such data can be simulated from the Weibull distribution of the survival times (times during which the patient/item is under observation till **the event or censoring**) with some exponentially distributed conditions on the occurance of the censoring.

The approach was taken from the StackOverflow question [How to create a toy survival (time to event) data with right censoring](https://stats.stackexchange.com/questions/135124/how-to-create-a-toy-survival-time-to-event-data-with-right-censoring) and is based on [Generating survival times to simulate Cox proportional hazards models](http://onlinelibrary.wiley.com/doi/10.1002/sim.2059/abstract)

``` r
library(survival)
set.seed(456)
x <- matrix(sample(0:1, size = 20000, replace = TRUE), ncol = 2)
head(x)
```

         [,1] [,2]
    [1,]    0    0
    [2,]    0    1
    [3,]    1    0
    [4,]    1    0
    [5,]    1    1
    [6,]    0    0

``` r
dCox <- dataCox(10^4, lambda = 3, rho = 2, x,
                beta = c(2,2), cens.rate = 5)
head(dCox)
```

      id       time status x.1 x.2
    1  1 0.04408843      0   0   0
    2  2 0.04919427      0   0   1
    3  3 0.17103077      0   1   0
    4  4 0.07568921      1   1   0
    5  5 0.05604885      1   1   1
    6  6 0.07529861      1   0   0

Fit the model
-------------

One can fit the model with the `coxphSGD()` function. Below is the explanation of parameters:

-   `formula` - the `formula` object passed in the same way you would with the `coxph` function
-   `data` - the list of data.frame's (containing the same columns) corresponding to batches of data
-   `epsilon` - the convergence parameter - optimization will stop when the difference of the estimates in subsequent steps will be smaller than epsilon (euclidean distance)
-   `learn.rates` - the function to be used to determine the step length in subsequent steps of the optimization process
-   `beta.zero` - vector containing start points for the optimization process
-   `max.iter` - the maximal number of algorithm iterations, when the iterations number exceeds the batches number then the data are used again (one full data usage / all batches usage is called an epoch)

``` r
batch_id <- sample(1:90, size = 10^4, replace = TRUE)
dCox_split <- split(dCox, batch_id)
results <-
  coxphSGD(formula     = Surv(time, status) ~ x.1+x.2,
           data        = dCox_split,
           epsilon     = 1e-5,
           learn.rates = function(x){1/(100*sqrt(x))},
           beta.zero   = c(0,0),
           max.iter    = 10*90)
```

Track estimates during each iteration on the contour lines plot
---------------------------------------------------------------

One can extract the estimaes of the Cox PH model for each of the iteration with the following code.

``` r
coeff_by_iteration <-
  as.data.frame(
    do.call(
      rbind,
      results$coefficients
      )
    )
head(coeff_by_iteration)
```

            x.1       x.2
    1 0.0000000 0.0000000
    2 0.1287670 0.1505911
    3 0.2077041 0.2279745
    4 0.2815383 0.2849962
    5 0.3466851 0.3416297
    6 0.3864788 0.4102790

Then having the estimates in each iteration of the optimization process, one can create a plot of the contour lines of the partial log-likelihood calculated on the full data to compare the paths of optimization process with the usage of **stochastic gradient descent order I** method to get the final Cox PH model estimates.

Here with a black triangle the original `beta.zero` (set to iteration 0) are marked and with a red square the point that was estimated by SGD I algorithm and with blue crossed dot the estimates calculated by the GD II (Newton-Raphson algorithm).

![](README_files/figure-markdown_github/unnamed-chunk-7-1.png)

### Used code

The code to reproduce the graph can be found in [this gist](https://gist.github.com/MarcinKosinski/909826b62f8849675f0980384fd6e28e).

Genomic data analysis
=====================

The Cancer Genome Atlas and RTCGA
---------------------------------

[The Cancer Genome Atlas](https://cancergenome.nih.gov/) data included in the [RTCGA](https://github.com/RTCGA/) family of R packages

> The Cancer Genome Atlas (TCGA) is a comprehensive and coordinated effort to accelerate our understanding of the molecular basis of cancer through the application of genome analysis technologies, including large-scale genome sequencing - <http://cancergenome.nih.gov/>.

<img src="https://raw.githubusercontent.com/RTCGA/RTCGA/master/RTCGA_workflow.png" height="500" width="800">

I have converted selected datasets from this study into a few separate packages that are hosted on Bioconductor. These R packages make selected datasets easier to access and manage. Datasets in **RTCGA** data packages are large and cover complex relations between clinical outcomes and genetic background.

Fit the complex model
---------------------

### Data overview

In the real life genetic problems for the algorithm application we will use the clinical outcomes (times and death status) for patients suffering from Breast Cancer and the information about genes that were mutated in the cancer tissue collected from the patient.

Data were processed during my master thesis and are stored in the [archivist](https://github.com/pbiecek/archivist) like repository and can be loaded with the following command

``` r
library(archivist)
testCox <- archivist::aread('MarcinKosinski/MasterThesis/3eebc99bd231b16a3ea4dbeec9ab5edb')
trainCox<- archivist::aread('MarcinKosinski/MasterThesis/1a06bef4a60a237bb65ca3e2f3f23515')
class(trainCox)
```

    [1] "list"

``` r
length(trainCox)
```

    [1] 98

``` r
head(
  trainCox[[1]][ # train data is a list of data frames that corresponds to batches
    c(7, 10),    # pick few rows
    c(210,302,356,898,911,1092:1093) # pick few columns, and survival outcomes
    ]
  )
```

         COL14A1 DOCK9 FASN SEMA5A SHPRH patient.vital_status times
    793        0     0    0      0     0                    0   268
    1068       1     0    0      1     0                    1  1171

### Actual fit

``` r
formulaSGD <- archivist::aread('MarcinKosinski/MasterThesis/064277e1c2a1fbea36d7d0ac518b9c8d')
# Surv(times, patient.vital_status ~ all posible genes
# how many genes are there in the whole formula?
length(
  strsplit(
    as.character(formulaSGD)[3],
    split = "+",
    fixed = T
  )[[1]]
)
```

    [1] 1091

One can evalute the code to fit the model or can download the already calculated model

``` r
coxphSGD(
  formula = formulaSGD,
  data = trainCox,
  learn.rates = function(x){1/x},
  max.iter = 490 # 98*5 / 5 full epoches
) -> model
```

``` r
model <- archivist::aread('MarcinKosinski/MasterThesis/446ac4dcb7d65bf39057bb341b296f1a')
```

### Coefficients/insights

Please find the top 20 absolute values of the coefficients from the model with the names of genes mutations they correspond to.

|          |    beta|  exp\_beta|
|----------|-------:|----------:|
| EGFR     |   0.977|      2.657|
| ATP2B2   |   0.723|      2.061|
| TP53     |   0.708|      2.030|
| NDST4    |   0.702|      2.018|
| CHRM2    |   0.642|      1.901|
| LRRC4C   |   0.624|      1.866|
| CDKN2A   |   0.615|      1.850|
| C3       |   0.594|      1.811|
| GLI2     |   0.593|      1.809|
| KIAA1409 |   0.575|      1.777|
| KEAP1    |   0.566|      1.762|
| Unknown  |   0.554|      1.741|
| TMEM132D |   0.537|      1.710|
| DNAH17   |  -0.540|      0.583|
| DLGAP2   |  -0.544|      0.580|
| BCORL1   |  -0.545|      0.580|
| PLXNC1   |  -0.570|      0.566|
| CDH1     |  -0.587|      0.556|
| FLRT2    |  -0.883|      0.413|
| BRAF     |  -0.954|      0.385|

Estimated surival curves
------------------------

One could see, from the previous table that the `BRAF` gene if mutated has the negative impact on the hazard ratio (hazard is decreased 0.385 times), which means that people with that mutation will live longer. In the same time the `EGFR` gene mutation has a positive impact on the hazard ratio (hazard is increased 2.657 times) so that people with that gene mutation has a high death rate and short survival times.

Let's see how the results of the Cox Proportional Hazards model where coefficients where estimated with the Stochastic Gradient Descent order I method corresponds to the trends in data. One visual way of testing the output is creating Kaplan-Meier estimates of survival curves for groups that corresponds to those genes mutations.

``` r
library(survminer)
fit_braf <- survfit(
  Surv(times, patient.vital_status) ~ BRAF,
  data = do.call(rbind, trainCox)
)
ggsurvplot(
  fit_braf,
  palette = c("#E7B800", "#2E9FDF"),
  surv.median.line = "hv",  # add the median survival pointer.
  legend.labs =             # change legend labels.
    c("No mutation in BRAF", "Mutation in BRAF"),    
  risk.table = TRUE,       # show risk table.
   pval = TRUE,             # show p-value of log-rank test.
   conf.int = TRUE,         # show confidence intervals for 
                            # point estimates of survival curves.
   xlim = c(0,2500),         # present narrower X axis, but not affect
                            # survival estimates.
   xlab = "Time in days",   # customize X axis label.
   break.time.by = 500,     # break X axis in time intervals by 500.
   ggtheme = theme_light(), # customize plot and risk table with a theme.
 risk.table.y.text.col = T, # colour risk table text annotations.
  risk.table.y.text = FALSE # show bars instead of names in text annotations
                            # in legend of risk table
)
```

![](README_files/figure-markdown_github/unnamed-chunk-13-1.png)

``` r
fit_egfr <- survfit(
  Surv(times, patient.vital_status) ~ EGFR,
  data = do.call(rbind, trainCox)
)
ggsurvplot(
  fit_egfr,
  palette = c("#E7B800", "#2E9FDF"),
  surv.median.line = "hv",  # add the median survival pointer.
  legend.labs =             # change legend labels.
    c("No mutation in EGFR", "Mutation in EGFR"),
  risk.table = TRUE,       # show risk table.
   pval = TRUE,             # show p-value of log-rank test.
   conf.int = TRUE,         # show confidence intervals for 
                            # point estimates of survival curves.
   xlim = c(0,2500),         # present narrower X axis, but not affect
                            # survival estimates.
   xlab = "Time in days",   # customize X axis label.
   break.time.by = 500,     # break X axis in time intervals by 500.
   ggtheme = theme_light(), # customize plot and risk table with a theme.
 risk.table.y.text.col = T, # colour risk table text annotations.
  risk.table.y.text = FALSE # show bars instead of names in text annotations
                            # in legend of risk table
)
```

![](README_files/figure-markdown_github/unnamed-chunk-13-2.png)

The code used for the whole analysis can be found [in this gist](https://gist.github.com/MarcinKosinski/bf8924c2ac1da613af675e66c6a12545)

Mathematical formulas
=====================

Cox model specification
-----------------------

Cox model assumptions
---------------------

Cox model (partial) log-likelihood function
-------------------------------------------

The Newton-Raphson method of optimization
-----------------------------------------

The Stochastic Gradient Descent (order I) method of optimization
----------------------------------------------------------------

Remarks for the analysis
------------------------

useR 2017
=========

-   [Long abstract](https://github.com/MarcinKosinski/useR2017/blob/master/abstract_MarcinKosinski.pdf)
-   [Short abstract](https://github.com/MarcinKosinski/useR2017/blob/master/abstract_MarcinKosinski_short.pdf)
-   [Presentation]()


NEWS.md
coxphSGD 0.2.0
==============

- 2017-07-04
- Version submitted to CRAN.
- Removed the `yihui/rlp` dependency and moved to regular `roxygen2` structure of the package's documentation.
- Removed dependencies (imports) of `assertthat`, `magrittr` and `reshape2` as it's better to have less dependencies in the
fastly evolving software development world (backward incompatibilities).
- Provided README.md
- Cleaned naming convention of functions and parameters.
- Created the package website.

coxphSGD 0.1.0
==============

- 2015-12-25 (released only on GitHub)
- Library was build with the experimental approach using `yihui/rlp` package
for creating package structures and documentation only in the vignette document.
- In the root directory of the package one could found the master thesis (in polish), 
entitled `Stochastic Gradient Descent Estimation in the Cox Proportional Hazards Model with
Applications to The Cancer Genome Atlas Data`. The package was the effect of tools written during
the study.
- Package was equipped with the [archivist](https://github.com/pbiecek/archivist) like
repository of R objects, where partial (and final results) of the analysis performed during
my master thesis were stored.

> The promotor of the master thesis was [profesor Przemyslaw Biecek](https://biecek.pl) (Warsaw University of Technology), who provided many improvements and comments to the package and who had a great impact on my master thesis and the whole scientific career, which I deeply appreciate


R/coxphSGD-GEN.R
#' Stochastic Gradient Descent log-likelihood estimation in
#' Cox proportional hazards model
#'
#' Function \code{coxphSGD} estimates coefficients using stochastic
#' gradient descent algorithm in Cox proportional hazards model.
#'
#' @param formula a formula object, with the response on the left of a ~ operator,
#' and the terms on the right. The response must be a survival object as returned by
#' the Surv function.
#' @param data a list of batch data.frames in which to interpret the variables named in the \code{formula}.
#' See Details.
#' @param learningRates a function specifing how to define learning rates in
#' steps of the algorithm. By default the \code{f(t)=1/t} is used, where \code{t} is
#' the number of algorithm's step.
#' @param beta_0 a numeric vector (if of length 1 then will be replicated) of length
#' equal to the number of variables after using \code{formula} in the \code{model.matrix}
#' function
#' @param epsilon a numeric value with the stop condition of the estimation algorithm.
#' @param max.iter numeric specifing maximal number of iterations.
#'
#' @details A \code{data} argument should be a list of data.frames, where in every batch data.frame
#' there is the same structure and naming convention for explanatory and survival (times, censoring)
#' variables. See Examples.
#'
#' @note If one of the conditions is fullfiled (j denotes the step number)
#' \itemize{
#'  \item \eqn{||\beta_{j+1}-\beta_{j}|| <}\code{epsilon} parameter for any \eqn{j}
#'  \item \eqn{j>max.iter}
#' }
#' the estimation process is stopped.
#' @export
#' @importFrom survival Surv
#' @importFrom assertthat assert_that
#' @importFrom magrittr %>%
#' @importFrom dplyr arrange
#' @importFrom dplyr filter
#' @importFrom stats model.extract
#' @importFrom stats rexp
#' @importFrom stats runif
#' @author
#' Marcin Kosinski, \email{m.p.kosinski@@gmail.com}
#' @examples
#' library(survival)
#' \dontrun{
#' coxphSGD(Surv(time, status) ~ ph.ecog + age,
#'          data= split(lung, sample(1:4, size = 228, replace = TRUE))
#' )
#' }
#'
coxphSGD <- function(formula, data, learningRates = function(x){1/x},
                    beta_0 = 0, epsilon = 1e-5, max.iter = 500 ) {
  # check arguments
  checkArguments(formula, data, learningRates,
                  beta_0, epsilon) -> beta_start
  n <- length(data)
  diff <- epsilon + 1
  i <- 1
  beta_new <- list()     # steps are saved in a list so that they can
  beta_old <- beta_start # be traced in the future
  # estimate
  while(i <= max.iter & diff > epsilon) {
    beta_new[[i]] <- coxphSGD_batch(formula = formula, beta = beta_old,
        learningRate = learningRates(i), data = data[[ifelse(i%%n==0,n,i%%n)]]) %>%
      unlist
    diff <- sqrt(sum((beta_new[[i]] - beta_old)^2))
    beta_old <- beta_new[[i]]
    i <- i + 1
    cat("\r iteration: ", i, "\r")
  }
  # return results
  list(Call = match.call(), epsilon = epsilon,
       learningRates = learningRates, steps = i,
       coefficients = c(list(beta_start), beta_new))
}

coxphSGD_batch <- function(formula, data, learningRate, beta){
  # collect times, status, variables and reorder samples
  # to make the algorithm more clear to read and track
  batchData <- prepareBatch(formula = formula, data = data)
  # calculate the log-likelihood for this batch sample
  batchData <- batchData %>% arrange(-times)
  # scores occure in nominator and denominator
  scores <- apply(batchData[, -c(1, 2)], 1,
                  function(element) exp(element %*% beta) )
  nominator <- apply(batchData[, -c(1, 2)], 2,
    						 function(element) cumsum(scores*element) )
  denominator <- cumsum(scores)
  # sum over non-censored observations
  partial_sum <- (batchData[, -c(1, 2)] - nominator/denominator)*batchData[, "event"]
  # each column indicates one explanatory variable
  U_batch <- partial_sum %>% colSums()
  return(beta + learningRate * U_batch)
}

checkArguments <- function(formula, data, learningRates,
                             beta_0, epsilon) {
  assert_that(is.list(data) & length(data) > 0)
  assert_that(length(unique(unlist(lapply(data, ncol)))) == 1)
  # + check names and types for every variables
  assert_that(is.function(learningRates))
  assert_that(is.numeric(epsilon))
  assert_that(is.numeric(beta_0))

    # check length of the start parameter
  if (length(beta_0) == 1) {
    beta_0 <- rep(beta_0, as.character(formula)[3] %>%
                    strsplit("\\+") %>%
                    unlist %>%
                    length)
  }

  return(beta_0)
}


prepareBatch <- function(formula, data) {
  # Parameter identification as in  `survival::coxph()`.
  Call <- match.call()
  indx <- match(c("formula", "data"),
                names(Call), nomatch = 0)
  if (indx[1] == 0)
      stop("A formula argument is required")
  temp <- Call[c(1, indx)]
  temp[[1]] <- as.name("model.frame")

  mf <- eval(temp, parent.frame())
  Y <- model.extract(mf, "response")

  if (!inherits(Y, "Surv"))
      stop("Response must be a survival object")
  type <- attr(Y, "type")

  if (type != "right" && type != "counting")
      stop(paste("Cox model doesn't support \"", type, "\" survival data",
          sep = ""))

  # collect times, status, variables and reorder samples
  # to make the algorithm more clear to read and track
  cbind(event = unclass(Y)[,2], # 1 indicates event, 0 indicates cens
        times = unclass(Y)[,1],
        mf[, -1]) %>%
    arrange(times)
}

#'
#' Cox Proportional Hazards Model Data Generation From Weibull Distribution
#'
#' Function \code{dataCox} generaters random survivaldata from Weibull
#' distribution (with parameters \code{lambda} and \code{rho} for given input
#' \code{x} data, model coefficients \code{beta} and censoring rate for censoring
#' that comes from exponential distribution with parameter \code{censRate}.
#'
#' @param N Number of observations to generate.
#' @param lambda lambda parameter for Weibull distribution.
#' @param rho rho parameter for Weibull distribution.
#' @param x A data.frame with input data to generate the survival times for.
#' @param beta True model coefficients.
#' @param censRate Parameter for exponential distribution, which is
#' responsible for censoring.
#'
#' @details For each observation true survival time is generated and a censroing time. If censoring time is less then survival time, then the survival time
#' is returned and a status of observations is set to \code{0} which means the
#' observation had censored time. If the survival time is less than censoring
#' time, then for this observation the true survival time is returned and the
#' status of this observation is set to \code{1} which means that the event has
#' been noticed.
#'
#' @references
#' \url{http://onlinelibrary.wiley.com/doi/10.1002/sim.2059/abstract}
#'
#' \code{Generating survival times to simulate Cox proportional hazards models}, 2005 by Ralf Bender, Thomas Augustin, Maria Blettner.
#' @importFrom reshape2 melt
#' @examples
#' \dontrun{
#' x <- matrix(sample(0:1, size = 20000, replace = TRUE), ncol = 2)
#' dataCox(10^4, lambda = 3, rho = 2, x,
#' beta = c(1,3), censRate = 5) -> dCox
#'}
#' @export
dataCox <- function(N, lambda, rho, x, beta, censRate){

  # real Weibull times
  u <- runif(N)
  Treal <- (- log(u) / (lambda * exp(x %*% beta)))^(1 / rho)

  # censoring times
  Censoring <- rexp(N, censRate)

  # follow-up times and event indicators
  time <- pmin(Treal, Censoring)
  status <- as.numeric(Treal <= Censoring)

  # data set
  data.frame(id=1:N, time=time, status=status, x=x)
}


# assuming observations are sorted by time
full_cox_loglik <- function(beta1, beta2, x1, x2, censored){
  sum(rev(censored)*(beta1*rev(x1) + beta2*rev(x2) -
                       log(cumsum(exp(beta1*rev(x1) + beta2*rev(x2))))))
}


calculate_outer_cox <- function(x1, x2, censored){
  ## contours
  outer_res <- outer(seq(-1,3, length = 100),
           seq(0,4, length = 100),
           Vectorize( function(beta1,beta2){
             full_cox_loglik(beta1, beta2, x1 = x1, x2 = x2, censored = censored)
           } )
  )
  outer_res_melted <- melt(outer_res)
  outer_res_melted$Var1 <- as.factor(outer_res_melted$Var1)
  levels(outer_res_melted$Var1) <- as.character(seq(-1,3, length = 100))
  outer_res_melted$Var2 <- as.factor(outer_res_melted$Var2)
  levels(outer_res_melted$Var2) <- as.character(seq(0,4, length = 100))
  outer_res_melted$Var1 <- as.numeric(as.character(outer_res_melted$Var1))
  outer_res_melted$Var2 <- as.numeric(as.character(outer_res_melted$Var2))
  return(outer_res_melted)
}


#'
#' Optimize Partial Log-Likelihood Function For Cox Propotional Hazards Model
#' Using Stochastic Gradient Descent
#'
#' Function \code{simulateCoxSGD} splits input \code{dCox} data on 10, 30, 60, 90, 120 and 200
#' groups and for each split it uses \link{coxphSGD} function to generate estimates for Cox
#' Proportional Hazards Model with stochastic gradient descent optimization.
#'
#' @param dCox Input data.frame containing survival times (columnd should be named \code{time}) and status
#' (columnd should be named \code{status}). So far this function only supports 2 explanatory variable
#' (that should be named \code{x1} and \code{x2}.
#' @param learningRates Parameter passed to \link{coxphSGD}.
#' @param epsilon Parameter passed to \link{coxphSGD}.
#' @param beta_0 Parameter passed to \link{coxphSGD}.
#' @param max.iter Parameter passed to \link{coxphSGD} to multiple the maximal iterations by the rows number.
#'
#' @importFrom dplyr bind_rows
#' @importFrom dplyr mutate
#' @examples
#' \dontrun{
#' x <- matrix(sample(0:1, size = 20000, replace = TRUE), ncol = 2)
#' dataCox(10^4, lambda = 3, rho = 2, x,
#' beta = c(1,3), censRate = 5) -> dCox
#'  simulateCoxSGD(dCox, learningRates = function(x){1/(100*sqrt(x))},
#'                 max.iter = 10, epsilon = 1e-5) -> d2ggplot
#'}
#' @export
simulateCoxSGD <- function(dCox = dCox, learningRates = function(x){1/x},
                      epsilon = 1e-03, beta_0 = c(0,0), max.iter = 100){

  # podziel na pozdzbiory
  sample(1:90, size = 10^4, replace = TRUE) -> group

  split(dCox, group) -> dCox_splitted
  coxphSGD(Surv(time, status)~x.1+x.2, data = dCox_splitted, epsilon = epsilon,
           learningRates = learningRates, beta_0 = beta_0, max.iter = max.iter*90) -> estimates



  sample(1:60, size = 10^4, replace = TRUE) -> group

  split(dCox, group) -> dCox_splitted


  coxphSGD(Surv(time, status)~x.1+x.2, data = dCox_splitted, epsilon = epsilon,
           learningRates = learningRates, beta_0 = beta_0, max.iter = max.iter*60) -> estimates2



  sample(1:120, size = 10^4, replace = TRUE) -> group

  split(dCox, group) -> dCox_splitted

  coxphSGD(Surv(time, status)~x.1+x.2, data = dCox_splitted, epsilon = epsilon,
           learningRates = learningRates, beta_0 = beta_0, max.iter = max.iter*120) -> estimates3


  sample(1:200, size = 10^4, replace = TRUE) -> group

  split(dCox, group) -> dCox_splitted

  coxphSGD(Surv(time, status)~x.1+x.2, data = dCox_splitted, epsilon = epsilon,
           learningRates = learningRates, beta_0 = beta_0, max.iter = max.iter*200) -> estimates4


  sample(1:30, size = 10^4, replace = TRUE) -> group

  split(dCox, group) -> dCox_splitted

  coxphSGD(Surv(time, status)~x.1+x.2, data = dCox_splitted, epsilon = epsilon,
           learningRates = learningRates, beta_0 = beta_0, max.iter = max.iter*30) -> estimates5


  t(simplify2array(estimates$coefficients)) %>%
    as.data.frame() -> df1
  t(simplify2array(estimates2$coefficients)) %>%
    as.data.frame() -> df2
  t(simplify2array(estimates3$coefficients)) %>%
    as.data.frame() -> df3
  t(simplify2array(estimates4$coefficients)) %>%
    as.data.frame() -> df4
  t(simplify2array(estimates5$coefficients)) %>%
    as.data.frame() -> df5

  df1 %>%
    mutate(version = paste("90 batches,", nrow(df1), " steps")) %>%
    bind_rows(df2 %>%
                mutate(version = paste("60 batches,", nrow(df2), " steps"))) %>%
    bind_rows(df3 %>%
                mutate(version = paste("120 batches,", nrow(df3), " steps"))) %>%
    bind_rows(df4 %>%
                mutate(version = paste("200 batches,", nrow(df4), " steps"))) %>%
    bind_rows(df5 %>%
                mutate(version = paste("30 batches,", nrow(df5), " steps")))  -> d2ggplot

  return(list(d2ggplot = d2ggplot, est1 = estimates, est2 = estimates2,
              est3 = estimates3, est4 = estimates4, est5 = estimates5))

}

